compiler/rustc_trait_selection/src/traits/project.rs

```diff
@@ -396,6 +415,55 @@ impl<'a, 'b, 'tcx> TypeFolder<'tcx> for AssocTypeNormalizer<'a, 'b, 'tcx> {
                 normalized_ty
             }
 
+            ty::Projection(data) if !data.trait_ref(self.tcx()).has_escaping_bound_vars() => {
+                // Okay, so you thought the previous branch was hacky. Well, to
+                // extend upon this, when the *trait ref* doesn't have escaping
+                // bound vars, but the associated item *does* (can only occur
+                // with GATs), then we might still be able to project the type.
+                // For this, we temporarily replace the bound vars with
+                // placeholders. Note though, that in the case that we still
+                // can't project for whatever reason (e.g. self type isn't
+                // known enough), we *can't* register an obligation and return
+                // an inference variable (since then that obligation would have
+                // bound vars and that's a can of worms). Instead, we just
+                // give up and fall back to pretending like we never tried!
+
+                let infcx = self.selcx.infcx();
+                let (data, mapped_regions, mapped_types, mapped_consts) =
+                    BoundVarReplacer::replace_bound_vars(infcx, &mut self.universes, data);
+                let normalized_ty = infcx
+                    .commit_if_ok(|_| {
+                        opt_normalize_projection_type(
+                            self.selcx,
+                            self.param_env,
+                            data,
+                            self.cause.clone(),
+                            self.depth,
+                            &mut self.obligations,
```

Comment:
Wait, something seems wrong here. Maybe I was wrong to suggest the `commit_if_ok`. 

In the event that `Ok(None)` is returned, I believe we already guarantee that `opt_normalize_projection_type` doesn't modify the inference state.

In the event that `Err` is returned, we will push things into `self.obligations`. If we're going to use `commit_if_ok`, we'd have to ensure that those things are removed, or else they are leaking through a transaction, and that's a bug.

I don't think it actually matters, though, it things *do* leak out in the case of an error. The only thing we push in there is an obligation to prove that the trait-ref holds, and it's ok for that to go on and be proven with the placeholders. This actually holds for any obligations that get pushed.

The case I was worried  about was something like: we return a type whose value will be informed b solving those obligations, but then we abandon it and use `ty` instead. But that just doesn't happen.